### PR TITLE
Update sdk and InspectCode version

### PR DESCRIPTION
**Pull request by huoyaoyuan**
This will unblock switch expressions, default interface implementations, and some other new features.

Notes for sdk version in `global.json`:
Previously, `global.json` doesn't support rolling forward, thus can only be used to lock the sdk version. From .NET Core 3, it can be used to specify _minimal required SDK version_ now. Reference: https://github.com/dotnet/designs/blob/master/accepted/global-json-updates.md
My opinions for setting rolling level:
- Prefer lower to higher. This helps finding usages requiring SDK version update
- Only rolling feature brands seems too strict. Will major rolling bring unpredictable changes? Minor should be a medium level.

Including index and range expressions in this PR as they are small changes.

NOT updating the target framework to 3.1. Should we do this now? Updating target framework will be much easier in .NET Core.

### CODE REVIEW

File: osu.Game.Tests/Visual/Online/TestSceneBeatmapSetOverlayDetails.cs [hunk 1]
```diff
@@ -44,7 +44,7 @@ public void TestMetrics()
             AddStep("set second set", () => details.BeatmapSet = secondSet);
             AddAssert("ratings set", () => details.Ratings.Metrics == secondSet.Metrics);
 
-            BeatmapSetInfo createSet() => new BeatmapSetInfo
+            static BeatmapSetInfo createSet() => new BeatmapSetInfo
```
Comment (huoyaoyuan) [inline osu.Game.Tests/Visual/Online/TestSceneBeatmapSetOverlayDetails.cs]: This is a fatal bug in VS16.3/roslyn3.3 and fixed now. Compiler is now bound to SDK version, forcing SDK 3.1.100 is required for this fix.

**Comment by huoyaoyuan** [comment]
It's a bit strange that my local InspectCode run catches some of index usages, but not all. Maybe due to cache issues.

**Comment by smoogipoo** [comment]
Welcome to JetBrains _still_ only having a primitive support of C#8.

Generally I’m ok with these changes. Just gonna ping @peppy to make sure we’re on the same page.

**Review by smoogipoo — APPROVED** [review]

File: osu.Game/Graphics/Containers/LinkFlowContainer.cs [hunk 2]
```diff
@@ -37,7 +37,7 @@ public void AddLinks(string text, List<Link> links)
 
             foreach (var link in links)
             {
-                AddText(text.Substring(previousLinkEnd, link.Index - previousLinkEnd));
+                AddText(text[previousLinkEnd..link.Index]);
```
Comment (peppy) [inline osu.Game/Graphics/Containers/LinkFlowContainer.cs]: am i alone in thinking this reads really badly?

Comment (smoogipoo) [reply osu.Game/Graphics/Containers/LinkFlowContainer.cs]: It's much safer this way, though. Very easy to get into off-by-one traps with index+length combos.

Comment (peppy) [reply osu.Game/Graphics/Containers/LinkFlowContainer.cs]: ok, let's go with it for now. looks alright with font ligatures.

**Pull request merged**